docs/develop/go/message-passing.mdx

```diff
@@ -192,309 +237,225 @@ if err != nil {
 
-```go
-var result string
-if err != response.Get(&result); err != nil {
-  // ...
-}
-log.Println("Received Query result. Result: " + result)
-```
+1. **Query failed.**
+    You'll get [`QueryFailed`](https://pkg.go.dev/go.temporal.io/api/serviceerror#QueryFailed).
+    Any panic in a Query handler will cause this.
+    Note that this differs from Signal and Update, for which panics can cause Workflow Task Failure.
 
-## Updates {#updates}
+1. **The handler caused the Workflow Task to fail.**
+    This would happen, for example, if the Query handler blocks the thread for too long without yielding.
 
-An [Update](/encyclopedia/workflow-message-passing#sending-updates) is an operation that can mutate the state of a Workflow Execution and return a response.
+## Message handler patterns {#message-handler-patterns}
 
-### Define an Update {#define-update}
+In addition to the topics below, see [Inject work into the main Workflow](/encyclopedia/workflow-message-passing#injecting-work-into-main-workflow) and [Ensuring your messages are processed exactly once](/encyclopedia/workflow-message-passing#exactly-once-message-processing).
 
-**How to define an Update using the Go SDK.**
+### Blocking handlers {#blocking-handlers}
 
-In Go, you define an Update type, also known as an Update name, as a `string` value.
-You must ensure the arguments and result are [serializable](/dataconversion).
-When sending and receiving the Update, use the Update name as an identifier.
-The name does not link to the data type(s) sent with the Update.
-Ensure that every Workflow listening to the same Update name can handle the same Update arguments.
+In the examples above, all the Update handlers did not block. But handlers can also be blocking. This means you can wait for Activities, Child Workflows, timers, `workflow.Await` conditions, etc, in Update Handlers thus opening up many powerful possibilities. However, this means that handler executions, and your main Workflow method, are all running concurrently, with switching occurring between them at await calls. (I.e. they interleave, but there is no parallelism.) It's essential to understand the things that could go wrong in order to use blocking handlers safely. See Workflow message passing for guidance on safe usage of async Update handlers, the Safe message handlers sample, and the Controlling handler concurrency and Waiting for message handlers to finish sections below.
 
-<div class="copycode-notice-container">
-  <a href="https://github.com/temporalio/documentation/blob/main/sample-apps/go/features/updates/your_updatable_workflow.go">
-    View the source code
-  </a>{' '}
-  in the context of the rest of the application code.
-</div>
+As an example of an blocking update handler, in the following code sample the Update handler from above has been changed to be an async def so that it can execute an Activity to make a network call to a remote service:
```

Comment:
```suggestion
As an example of an blocking update handler, in the following code sample the Update handler from above has been changed so that it can execute an Activity to make a network call to a remote service:
```